lib/Diagnostics/Diagnostics.js

```diff
@@ -2,35 +2,119 @@
+    a boolean flag for the class to know if data needs to be updated
+    to server.
+    StaticMetaData = {
+        modified: false
+        data: {},
+    }
+    2) An object called DynamicMetaData containing two arrays.
+    Dynamic object is sent to server regularly
+    DynamicMetaData = {
+        inputs: [],
+        flavors: []
+    }
+ */
+/********************************************************************/
+
+const StreamHealth = {
+  Good: 'Good',
+  Fair: 'Fair',
+  Poor: 'Poor'
+};
+
+class Diagnostics extends EventEmitter {
+    constructor(entryObject) {
+        super();
+        this.entryId = entryObject.entryId;
+        this.partnerId = entryObject.partnerId;
+        this._diagnosticsEnd = Q.defer();
+        this._shouldRun = true;
+        this._logger = loggerModule.getLogger("Diagnostics", `[${entryObject.entryId}] `);
+        this._syncPromise = new PromUtils.SynchornizedPromises(this._logger);
+        this._reportIntervalTimeInMSec = config.get('diagnostics').diagnosticsReportIntervalRunTimeInSec *1000;
         this.alertWindowSizeInMSec = config.get('diagnostics').errorsWindowSizeInMin * 60 * 1000;
+        this.lastDiagnosticsObjTime = 0;
+        this.staticMetaData = { modified: false };
+        this.dynamicMetaData = { inputs: [], flavors: [] };
+        this.reportedAlerts = new Map();
+        this._initAlertsAnalyzer.call(this, entryObject.entryId);
+
         this._internalAlerts = [];
-        this._externalAlerts = new Map();
-        this._logger = loggerModule.getLogger("Diagnostics", `[${loggerInfo}] `);
-        this.clearOldAlerts.call(this);
+        this._runDiagnosticsReport.call(this);
     }
 
     toJSON() {
-        return { "internalAlerts" : this._internalAlerts, "externalAlerts" : Array.from(this._externalAlerts) };
+        return { "internalAlerts" : this._internalAlerts, "reportedAlerts" : Array.from(this.reportedAlerts) };
     }
-    
-    pushAlert(diagnosticsAlertObj) {
+
+    _initAlertsAnalyzer(entryId) {
+        this._diagnosticsAnalyzer = new DiagnosticsAlertsAnalyzer(entryId, this.alertWindowSizeInMSec);
+        // Enlist on events
+        this._diagnosticsAnalyzer.on('insertAlert', this.insertAlert.bind(this));
+        this._diagnosticsAnalyzer.on('removeAlert', this.removeAlert.bind(this));
+    }
+
+    cleanup() {
+        this._logger.debug('Stopping diagnostics module');
+        this._shouldRun = false;
+        return this._diagnosticsEnd.promise
+            .then(()=> {
+                this._logger.debug('Diagnostics Module stopped');
+            })
+    }
+
+    insertAlert(diagnosticsAlertObj) {
+        this._logger.debug(`[insertAlert] Alert: ${diagnosticsAlertObj.name}`);
         if (diagnosticsAlertObj.hashKey) {
-            this._externalAlerts.set(diagnosticsAlertObj.hashKey, diagnosticsAlertObj);
+            this.reportedAlerts.set(diagnosticsAlertObj.hashKey, diagnosticsAlertObj);
+            return this._updateStreamHealth.call(this)
```

Comment:
won't it be better to mark a flag as dirty and report only once at the end of all alerts calculations?